# *data load*

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('../../data/sentiment.csv')
df = df[['content', 'vec', 'sentiment']]

In [3]:
df.head()

content  \
0  G7 원전사고 국경 넘어 민간인에 방사성 물질 노출 위험 무력분쟁 핵시설 파손으로 ...   
1  에 꽂힌 핀다   렌트리스 이어 오토론까지 보폭 확대 핀다 정상연 수석매니저 인터뷰...   
2  3000명 회원에게 제각기 다른 3000개 앱이 되겠다여성 패션앱 1위 에이블리의 ...   
3  항공업계 러시아 운항영공 통과 중단우회항로 이용 기사내용 요약대한항공아시아나 등 우...   
4  에 꽂힌 핀다  렌트리스 이어 오토론까지 보폭 확대 핀다 정상연 수석매니저 인터뷰2...   

                                                 vec  sentiment  
0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2  
1  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0451434348289...          1  
2  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1  
3  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2  
4  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0450219082043...          1

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5011 entries, 0 to 5010
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   content    5011 non-null   object
 1   vec        5011 non-null   object
 2   sentiment  5011 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 117.6+ KB


# *make train, test*

In [5]:
from sklearn.model_selection import train_test_split

In [6]:
train_data, test_data = train_test_split(df, test_size=0.2, random_state=42)

In [7]:
train_data = train_data.reset_index()
test_data = test_data.reset_index()

In [8]:
print(len(df))
print(len(train_data))
print(len(test_data))

5011
4008
1003


In [9]:
# test
test_vec = []
for i in test_data.vec:
    test_vec.append(eval(i))


test_label = [i for i in test_data.sentiment]

In [10]:
# train
train_vec = []
for i in train_data.vec:
    train_vec.append(eval(i))


train_label = [i for i in train_data.sentiment]

In [11]:
train_vec[1][:10]

[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

In [12]:
print(len(train_vec))
print(len(test_vec))

4008
1003


# *classifier*

In [13]:
# kNN
from sklearn.neighbors import KNeighborsClassifier
# Decision Tree
from sklearn.tree import DecisionTreeClassifier
# Random Forest
from sklearn.ensemble import RandomForestClassifier
# 나이브 베이즈
from sklearn.naive_bayes import GaussianNB
# 서포터 벡터 머신
from sklearn.svm import SVC

In [14]:
from sklearn.model_selection import KFold, cross_val_score

In [15]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   index      4008 non-null   int64 
 1   content    4008 non-null   object
 2   vec        4008 non-null   object
 3   sentiment  4008 non-null   int64 
dtypes: int64(2), object(2)
memory usage: 125.4+ KB


### *score*

In [16]:
scoring = 'accuracy'

In [17]:
k_fold = KFold(shuffle = True, random_state=0)

In [ ]:
# knn
knn_clf = KNeighborsClassifier()
knn_clf = knn_clf.fit(train_vec, train_label)
knn_score = cross_val_score(knn_clf, train_vec, train_label, cv=k_fold, n_jobs=1, scoring=scoring)

# decision tree
dt_clf = DecisionTreeClassifier()
dt_clf = dt_clf.fit(train_vec, train_label)
dt_score = cross_val_score(dt_clf, train_vec, train_label, cv=k_fold, n_jobs=1, scoring=scoring)

# random forest
rf_clf = RandomForestClassifier()
rf_clf = rf_clf.fit(train_vec, train_label)
rf_score = cross_val_score(rf_clf, train_vec, train_label, cv=k_fold, n_jobs=1, scoring=scoring)

# naive bayes
nb_clf = GaussianNB()
nb_clf = nb_clf.fit(train_vec, train_label)
nb_score = cross_val_score(nb_clf, train_vec, train_label, cv=k_fold, n_jobs=1, scoring=scoring)

# svm
svm_clf = SVC()
svm_clf = svm_clf.fit(train_vec, train_label)
svm_score = cross_val_score(svm_clf, train_vec, train_label, cv=k_fold, n_jobs=1, scoring=scoring)

In [ ]:
svm_score

In [ ]:
print(f"kNN_score : {round(np.mean(knn_score) * 100, 2)}")
print(f"DecisionTree_score : {round(np.mean(dt_score) * 100, 2)}")
print(f"RandomForest_score : {round(np.mean(rf_score) * 100, 2)}")
print(f"naivebayes_score : {round(np.mean(nb_score) * 100, 2)}")
print(f"svm_score : {round(np.mean(svm_score) * 100, 2)}")

# *doc2vec*
##### *vector_size 256*
##### *kNN_score : 64.52*
##### *DecisionTree_score : 56.52*
##### *RandomForest_score : 71.9*
##### *naivebayes_score : 53.9*
##### *svm_score : 70.54*
# 
# *TF-IDF*
##### *kNN_score : 75.27*
##### *DecisionTree_score : 74.08*
##### *RandomForest_score : 82.81*
##### *naivebayes_score : 76.52*
##### *svm_score : 82.26*

# 

### *Random Forest*

In [35]:
from sklearn.ensemble import RandomForestClassifier

In [36]:
rf_clf = RandomForestClassifier()

In [37]:
rf_clf.fit(train_vec, train_label)

RandomForestClassifier()

In [38]:
test_pred = rf_clf.predict(test_vec)

In [39]:
# labeling
test_data['label'] = test_pred

In [40]:
test_data.head(15)

index                                            content  \
0    1052  뤼이드 세계서 가장 혁신적인 기업 선정 뤼이드는 미국 기술 경제 전문지 패스트컴퍼니...   
1    3074  정은경 면역저하제 대상 이부실드 물량예산 확보 검토 방역당국이 아스트라제네카의 면역...   
2    1789  삼성전자 2022년형 네오 QLED TV 14일까지 사전 판매 개시 삼성전자가 20...   
3     764  LGU 서강대와 메타버스 대학 만든다 LG유플러스는 서강대학교와 메타버스 유니버시티...   
4    3822  세포막 모사한 초정밀 바이오센서 상용화 길 열었다 세포막은 세포 안팎의 보호막이자 ...   
5    1101  구글 54억 달러에 사이버보안 업체 맨디언트 인수 구글은 8일 사이버보안 업체 맨디...   
6    1666  티몬 최고기술책임자로 황태현 CTO 선임 티몬은 최고기술책임자로 황태현 전 구글 검...   
7     279  여성 위암 남성 보다 진단 어렵고 사망 원인도 달라 성별에 따라 위암의 특성과 사망...   
8    3541  무빙 전암 연구소장 이륜차 매연소음 해결하고파 지난해 국내 배달 시장 규모는 약 2...   
9    4382  오영호 전 산업자원부 제1차관 별세 오영호 전 산업자원부 제1차관오영호 전 산업자원...   
10   2714  오미크론 증상발현 후 8일간 바이러스 배출7일 격리 후 조심해야 오미크론 변이바이러...   
11   4283  작년 무선이어폰 출하량 3억대   애플 에어팟 1위 작년 무선이어폰 세계 출하량이 ...   
12   2290  카모아 누적 거래대금 580억원 MZ세대가 주고객 렌트카 가격 비교 1등 앱 카모아...   
13   3288  신성이엔지 삼성 평택 투자 수혜 신성이엔지가 삼성전자의 경기 평택캠퍼스 투자 수혜를...   
14   4877  크롬 개인정보 보호하면서 맞춤형 광고 지원하는 제3의 길 제시 맞춤형 광고 업계가 ...   

                                                  vec  sentiment  label  
0   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
1   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      2  
2   [0.0, 0.0, 0.0, 0.0, 0.04822845993011052, 0.0,...          1      1  
3   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
4   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
5   [0.0, 0.0, 0.0, 0.0, 0.07299524458843958, 0.0,...          2      1  
6   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
7   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      2  
8   [0.0, 0.0, 0.0, 0.0, 0.025147560831665295, 0.0...          1      1  
9   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      2  
10  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      2  
11  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
12  [0.0, 0.0, 0.0, 0.0, 0.031637613238312524, 0.0...          1      1  
13  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
14  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      1

In [41]:
test_data.label.value_counts()

1    638
2    347
0     18
Name: label, dtype: int64

In [42]:
from sklearn.metrics import f1_score
f1_score(test_data.sentiment, test_pred, average='micro')

0.8354935194416749

In [51]:
test_data[test_data['label']==0]

index                                            content  \
41    2843  의료 데이터 기업 줄까 말까 문재인 정부는 정권 초부터 보건의료 빅데이터의 잠재가치...   
109   4690  예금보험공사 우리금융 지분 2 2 매각 예금보험공사가 11일 주식 시장 개장 전 우...   
144   4455  지금이 원격의료 도입을 위한 골든타임 혹자는 미래의료가 정보통신 기술과 결합해 개인...   
178   4594  자가검사키트 중고나라당근마켓 판매구매는 불법 다음달 5일까지 코로나19 자가검사키트...   
240   4675  CDC 오피오이드 진통제 처방에 대한 지침 변경 제안 미국 최고의 공중 보건 기관은...   
242   4160  루카셴코 푸틴 어느 때보다 건강소련 붕괴는 비극 알렉산드르 루카셴코 벨라루스 대통령...   
282   4094  아이폰SE3 1 8m 높이서 떨어뜨렸더니아이폰13과 내구성 비슷 애플이 최근 공개한...   
319   2211  코로나19 널리 퍼질수록 위중증 환자는 감소 확산이 빠르지만 위중증 발전 위험은 낮...   
358   2232  한국게임학회 확률형아이템 정보 공개 포함한 게임법 통과 요구 한국게임학회은 11일 ...   
419   2072  3월 1일부터 확진자 동거인 의무격리 안한다수동감시 전환 다음달 1일부터 확진자의 ...   
482   4521  기업 비트코인 투자한다면 522 자산 배분이 적정 법인의 암호화폐 투자에 대해 적정...   
498   2118  복지부 추경예산 5636억원 확정 올해 제1차 보건복지부 추가경정예산이 5천636억...   
661   2251  코로나 N만명 시대   IT기업 원격근무 강화 오미크론 변이 대유행으로 코로나19 ...   
694   4228  우크라러 4차 휴전 협상 16일 재개  타협 여지 있어 우크라이나와 러시아의 4차 ...   
730   2095  벤처기업 제외 업종에 가상자산 사업자 빠져야 가상자산 거래소를 대변하는 한국디지털자...   
962   4617  환경부 절수설비 절수등급 표시 의무화 앞으로 수입하는 절수설비에 절수등급 표시가 의...   
964   2629  유행 정체는 분명해도 감소세 전환은 모른다 방역당국이 오미크론 변이바이러스 유행 정...   
994   4453  연이은 횡령 사고에 웹케시 브랜치 4 0 도입 문의 증가 최근 기업  기관에서 연이...   

                                                   vec  sentiment  label  
41   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
109  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
144  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
178  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
240  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
242  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
282  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      0  
319  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
358  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
419  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
482  [0.0, 0.0, 0.0, 0.0, 0.02325799698661835, 0.0,...          0      0  
498  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
661  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
694  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
730  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
962  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      0  
964  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      0  
994  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      0

### *metric*

In [24]:
from sklearn.metrics import classification_report, f1_score, accuracy_score

##### *GridSearchCV*

In [25]:
from sklearn.model_selection import GridSearchCV

In [27]:
params = { 'n_estimators' : [100, 200],
           'max_depth' : [6, 8, 10, 12],
           'min_samples_leaf' : [8, 12, 18],
           'min_samples_split' : [8, 16, 20]
            }

# RandomForestClassifier 객체 생성 후 GridSearchCV 수행
rf_clf = RandomForestClassifier(random_state = 0, n_jobs = -1)
grid_cv = GridSearchCV(rf_clf,
                       param_grid = params,
                       cv = 3,
                       n_jobs = -1)
grid_cv.fit(train_vec, train_label)

print('최적 하이퍼 파라미터: ', grid_cv.best_params_)
print('최고 예측 정확도: {:.4f}'.format(grid_cv.best_score_))

최적 하이퍼 파라미터:  {'max_depth': 12, 'min_samples_leaf': 8, 'min_samples_split': 8, 'n_estimators': 200}
최고 예측 정확도: 0.7116


##### *최적 하이퍼 파라미터로 재학습*

In [28]:
rf_clf = RandomForestClassifier(n_estimators = 200,          # 결정 트리 개수
                                 max_depth = 12,            # 트리의 최대 깊이
                                 min_samples_leaf = 8,      # 리프노드가 되기 위해 필요한 최소한의 샘플 데이터수 
                                 min_samples_split = 8,    # 노드를 분할하기 위한 최소한의 샘플 데이터수(과적합을 제어하는데 사용)
                                 random_state = 0,
                                 n_jobs = -1)

In [29]:
rf_clf.fit(train_vec, train_label)

RandomForestClassifier(max_depth=12, min_samples_leaf=8, min_samples_split=8,
                       n_estimators=200, n_jobs=-1, random_state=0)

In [30]:
test_pred = rf_clf.predict(test_vec)

In [31]:
# labeling
test_data['label'] = test_pred

In [32]:
test_data.head(15)

index                                            content  \
0    1052  뤼이드 세계서 가장 혁신적인 기업 선정 뤼이드는 미국 기술 경제 전문지 패스트컴퍼니...   
1    3074  정은경 면역저하제 대상 이부실드 물량예산 확보 검토 방역당국이 아스트라제네카의 면역...   
2    1789  삼성전자 2022년형 네오 QLED TV 14일까지 사전 판매 개시 삼성전자가 20...   
3     764  LGU 서강대와 메타버스 대학 만든다 LG유플러스는 서강대학교와 메타버스 유니버시티...   
4    3822  세포막 모사한 초정밀 바이오센서 상용화 길 열었다 세포막은 세포 안팎의 보호막이자 ...   
5    1101  구글 54억 달러에 사이버보안 업체 맨디언트 인수 구글은 8일 사이버보안 업체 맨디...   
6    1666  티몬 최고기술책임자로 황태현 CTO 선임 티몬은 최고기술책임자로 황태현 전 구글 검...   
7     279  여성 위암 남성 보다 진단 어렵고 사망 원인도 달라 성별에 따라 위암의 특성과 사망...   
8    3541  무빙 전암 연구소장 이륜차 매연소음 해결하고파 지난해 국내 배달 시장 규모는 약 2...   
9    4382  오영호 전 산업자원부 제1차관 별세 오영호 전 산업자원부 제1차관오영호 전 산업자원...   
10   2714  오미크론 증상발현 후 8일간 바이러스 배출7일 격리 후 조심해야 오미크론 변이바이러...   
11   4283  작년 무선이어폰 출하량 3억대   애플 에어팟 1위 작년 무선이어폰 세계 출하량이 ...   
12   2290  카모아 누적 거래대금 580억원 MZ세대가 주고객 렌트카 가격 비교 1등 앱 카모아...   
13   3288  신성이엔지 삼성 평택 투자 수혜 신성이엔지가 삼성전자의 경기 평택캠퍼스 투자 수혜를...   
14   4877  크롬 개인정보 보호하면서 맞춤형 광고 지원하는 제3의 길 제시 맞춤형 광고 업계가 ...   

                                                  vec  sentiment  label  
0   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
1   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          0      1  
2   [0.0, 0.0, 0.0, 0.0, 0.04822845993011052, 0.0,...          1      1  
3   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
4   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
5   [0.0, 0.0, 0.0, 0.0, 0.07299524458843958, 0.0,...          2      1  
6   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
7   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      1  
8   [0.0, 0.0, 0.0, 0.0, 0.025147560831665295, 0.0...          1      1  
9   [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      1  
10  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      2  
11  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
12  [0.0, 0.0, 0.0, 0.0, 0.031637613238312524, 0.0...          1      1  
13  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          1      1  
14  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...          2      1

In [33]:
test_data.label.value_counts()

1    818
2    185
Name: label, dtype: int64

In [34]:
f1_score(test_data.sentiment, test_pred, average='micro')

0.7248255234297108

##### *classification_report*

In [52]:
classification_report(test_data.sentiment.tolist(), test_pred)

'              precision    recall  f1-score   support\n\n           0       0.56      0.10      0.18        96\n           1       0.84      0.95      0.89       568\n           2       0.83      0.85      0.84       339\n\n    accuracy                           0.84      1003\n   macro avg       0.74      0.64      0.64      1003\nweighted avg       0.81      0.84      0.81      1003\n'

              precision    recall  f1-score   support
0                 0.56      0.10      0.18        96
1                 0.84      0.95      0.89       568
2                 0.83      0.85      0.84       339
accuracy                              0.84      1003
macro avg         0.74      0.64      0.64      1003
weighted avg       0.81      0.84      0.81      1003

##### *f1_score*

In [53]:
f1_score(test_data.sentiment, test_pred, average='micro')

# 기본 random forest
# 0.8344965104685942

0.8354935194416749

##### *accuracy*

In [54]:
accuracy_score(test_data.sentiment, test_pred)

# 기본 random forest
# 0.8344965104685942

0.8354935194416749

### *SVM*

In [20]:
# svm_clf.fit(train_vec, train_label)

In [26]:
# score = svm_clf.predict(test_vec)

In [21]:
# test_data['label'] = score

In [22]:
# test_data.head(15)